sway/desktop/render.c

```diff
@@ -1394,23 +1721,47 @@ void output_render(struct sway_output *output, struct timespec *when,
 		fullscreen_con = workspace->current.fullscreen;
 	}
 
-	fx_renderer_begin(renderer, wlr_output->width, wlr_output->height);
-
 	if (debug.damage == DAMAGE_RERENDER) {
-		int width, height;
-		wlr_output_transformed_resolution(wlr_output, &width, &height);
 		pixman_region32_union_rect(damage, damage, 0, 0, width, height);
 	}
 
-	if (!pixman_region32_not_empty(damage)) {
-		// Output isn't damaged but needs buffer swap
-		goto renderer_end;
+
+	bool blur_optimize_should_render = false;
+	bool damage_not_empty = pixman_region32_not_empty(damage);
+	pixman_region32_t extended_damage;
+	pixman_region32_init(&extended_damage);
+	if (!fullscreen_con && !server.session_lock.locked && damage_not_empty) {
+		// Check if there are any windows to blur
+		if (should_blur_draw_optimize(output)) {
+			blur_optimize_should_render = true;
+			// Damage the whole output
+			pixman_region32_union_rect(damage, damage, 0, 0, width, height);
+		}
+
+		// Extend the damaged region
+		int expanded_size = get_container_expanded_size(NULL);
```

Comment:
This is the one area where that function is called, and currently could not work on some systems. It's possible to have blur disabled globally, but enabled on some containers (same with shadows). May want to rethink how we do this

Reply:
We could iterate through the windows and check each container just like `should_blur_draw_optimize`. It would be more efficient to check if any container wants blur/shadows at the same time as checking if we should render optimized blur